File: Operators/search_de_current_best.m
function [output1,output2]  = search_de_current_best(varargin)
% The "current-to-best/1" differental mutation.

mode = varargin{end};
switch mode
    case 'execute'
        Parent = varargin{1};
        Para   = varargin{3};

        F        = Para(1);
        CR       = Para(2);
        [~,best] = min(Parent.fits);
        GbestDec = Parent(best).dec; % the best solution
        Parent   = Parent.decs;

        [N,D]   = size(Parent);
        Parent1 = Parent;
        Parent2 = repmat(GbestDec,N,1);
        Parent3 = Parent1(randperm(N),:);

        ind = rand(N,D) < CR;
        Offspring = Parent;
        Offspring(ind) = Parent1(ind) + F*(Parent2(ind)-Parent3(ind));
        output1 = Offspring;
        output2 = varargin{5};

    case 'parameter'
        output1 = [0,1;0,1]; % F and CR

    case 'behavior'
        output1 = {'LS','small','small';'GS','large','large'}; % small F and CR values perform local search
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: AutoOptLib for Python/components/search_de_current_best.py
"""Python translation of search_de_current_best.""" 

from __future__ import annotations

import numpy as np


def search_de_current_best(*args):
    mode = args[-1]
    if mode == "execute":
        parent_obj = args[0]
        para = args[2] if len(args) > 2 else None
        aux = args[3] if len(args) > 3 else None  # noqa: F841
        fits = getattr(parent_obj, "fits", None)
        if callable(fits):
            fit_vals = np.asarray(fits()).reshape(-1)
        else:
            fit_vals = np.asarray(fits if fits is not None else getattr(parent_obj, "fit", None)).reshape(-1)
        best_idx = int(np.argmin(fit_vals))
        best_dec = getattr(parent_obj[best_idx], "dec", None)
        if callable(best_dec):
            gbest = np.asarray(best_dec())
        else:
            gbest = np.asarray(best_dec if best_dec is not None else getattr(parent_obj, "decs")[best_idx])
        decs = getattr(parent_obj, "decs", None)
        if callable(decs):
            parent = decs()
        else:
            parent = decs if decs is not None else parent_obj
        parent = np.asarray(parent, dtype=float)
        n, d = parent.shape
        if para is None:
            f, cr = 0.5, 0.5
        else:
            arr = np.asarray(para).reshape(-1)
            f = float(arr[0])
            cr = float(arr[1]) if arr.size > 1 else 0.5
        p1 = parent.copy()
        p2 = np.repeat(gbest.reshape(1, -1), n, axis=0)
        p3 = parent[np.random.permutation(n)]
        mask = np.random.rand(n, d) < cr
        offspring = parent.copy()
        offspring[mask] = p1[mask] + f * (p2[mask] - p3[mask])
        return offspring, args[3] if len(args) > 3 else None
    if mode == "parameter":
        return [[0, 1], [0, 1]], None
    if mode == "behavior":
        return [["LS", "small", "small"], ["GS", "large", "large"]], None
    raise ValueError(f"Unsupported mode: {mode}")
